cloud-init/qemu_performance.py

```diff
@@ -0,0 +1,549 @@
+     6. grep update /var/log/cloud-init.log to assert whether apt-get update
+        is called in default cases.
+
+    Use mount-image-callback to inject deb_path into a derivative daily image.
+    Boot original daily image and report boottime stats.
+    Boot derivative image and report boottime stats.
+    """
+    orig_data = []
+    new_data = []
+    existing_data_files = [f for f in data_dir.rglob("*.json")]
+    if existing_data_files:
+        print(f"--- Using pre-existing data files in {data_dir}")
+        daily = "pre-existing samples"
+        for f in existing_data_files:
+            if "avg" in f.name:
+                continue
+            if "orig" in f.name:
+                orig_data.append(json.loads(f.read_text()))
+            else:
+                new_data.append(json.loads(f.read_text()))
+    else:
+        data_dir.mkdir(exist_ok=True)
+        with pycloudlib.Qemu(tag="examples") as cloud:
+            daily = cloud.daily_image(release=release)
+            print(
+                f"--- Creating modified daily image {daily} with cloud-init"
+                f" from {deb_path}"
+            )
+            new_image = update_cloud_init_in_img(daily, deb_path, suffix=1)
+            print(
+                f"--- Launching {sample_count} control daily images {daily} ---"
+            )
+            for sample in range(sample_count):
+                instance = cloud.launch(image_id=daily, user_data=user_data)
+                orig_data.append(collect_bootspeed_data(instance))
+                instance.delete()
+                new_instance = cloud.launch(image_id=new_image, user_data=user_data)
+                new_data.append(collect_bootspeed_data(new_instance))
+                new_instance.delete()
+    orig_avg = update_averages(orig_data, data_dir, "orig")
+    new_avg = update_averages(new_data, data_dir, "new")
+    print(
+        HEADER.format(
+            boot_serial=orig_data[0]["image_builddate"],
+            deb_path=deb_path,
+            source_image=str(daily),
+            sample_count=sample_count,
+            min_avg_delta=MIN_AVG_PERCENT_DELTA,
+        )
+    )
+    report_errors(new_data)
+    inspect_averages(orig_avg, new_avg)
+
+
+CLOUDINIT_ERRORS = """\
+------------ Unexpected errors  ---------------------------------------
+Sample name: new-instance-{idx}-artifacts.json
+Errors:
+{errors}
+Recoverable Errors:
+{recoverable_errors}
+-----------------------------------------------------------------------
+"""
+
+
+def report_errors(samples: list):
+    """Print any unexpected errors found data samples"""
+    for idx, sample in enumerate(samples):
+        status = sample["cloudinit_status"]
+        if status["errors"] or status["recoverable_errors"]:
+            print(CLOUDINIT_ERRORS.format(idx=idx, **status))
+
+
+def get_parser():
+    parser = ArgumentParser(description=__doc__)
+    parser.add_argument(
+        "-n",
+        "--number-of-launches",
```

Comment:
The UI and logic that this represents could just as easily be the [pytest-repeat](https://pypi.org/project/pytest-repeat/) plugin if this script were part of our test infrastructure.